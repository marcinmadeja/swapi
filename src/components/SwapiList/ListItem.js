import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { swapiService } from 'services';

import { withStyles } from 'material-ui/styles';
import Card, { CardContent, CardActions } from 'material-ui/Card';
import IconButton from 'material-ui/IconButton';
import FavoriteIcon from 'material-ui-icons/Favorite';
import LinkIcon from 'material-ui-icons/Link';
import {
  styles,
} from './ListItem.styles';

export const createSeName = (name, url, listName) => {
  const seoName = swapiService.formatUrlName(name);
  const filmId = swapiService.getIdFromLink(url);
  return `/${listName}/${seoName}/${filmId}`;
};

const ListItem = ({
  classes = {},
  ListDetails,
  listName,
  item,
  item: {
    name,
    url,
  },
}) => {
  const itemURL = createSeName(name, url, listName);

  return (
    <Card className={classes.card}>
      <CardContent className={classes.cardContent}>
        <ListDetails details={item} />

        <CardActions className={classes.cardActions}>
          <IconButton aria-label="Add to favorites">
            <FavoriteIcon />
          </IconButton>

          <IconButton component={props => <Link to={itemURL} {...props} />} aria-label="Link">
            <LinkIcon />
          </IconButton>
        </CardActions>
      </CardContent>
    </Card>
  );
};

ListItem.propTypes = {
  classes: PropTypes.object.isRequired,
  ListDetails: PropTypes.func.isRequired,
  listName: PropTypes.string.isRequired,
  item: PropTypes.shape({
    name: PropTypes.string,
    url: PropTypes.string,
  }).isRequired,
};

export default withStyles(styles)(ListItem);
