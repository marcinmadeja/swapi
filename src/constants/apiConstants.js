export const SWAPI_URL = 'https://swapi.co/api';
export const SWAPI_FILMS = '/films/?format=json';

export const RANDOM_USERS = 'https://randomuser.me/api';
