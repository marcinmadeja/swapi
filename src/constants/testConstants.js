export const users = [
  {
    gender: 'male',
    name: {
      title: 'mr',
      first: 'terrence',
      last: 'watts',
    },
    location: {
      street: '5796 rolling green rd',
      city: 'bowral',
      state: 'victoria',
      postcode: 7762,
      coordinates: {
        latitude: '-79.9438',
        longitude: '-160.0279',
      },
      timezone: {
        offset: '-7:00',
        description: 'Mountain Time (US & Canada)',
      },
    },
    email: 'terrence.watts@example.com',
    login: {
      uuid: 'ef30b2c2-7575-42d1-b9e5-38030d64b184',
      username: 'tinygoose275',
      password: 'makaveli',
      salt: 'EwEwiYhD',
      md5: '67600680d1f62c990021d597e96b264d',
      sha1: '174b563d3f42d98f88e7bcaad064e647ee726faa',
      sha256: 'fc93dd24693ad0650413b01980ccc494dd3921991a8ab41e45cadf8610aa046a',
    },
    dob: {
      date: '1972-09-08T12:37:28Z',
      age: 45,
    },
    registered: {
      date: '2003-03-25T20:55:50Z',
      age: 15,
    },
    phone: '00-8437-1526',
    cell: '0476-938-903',
    id: {
      name: 'TFN',
      value: '686111438',
    },
    picture: {
      large: 'https://randomuser.me/api/portraits/men/24.jpg',
      medium: 'https://randomuser.me/api/portraits/med/men/24.jpg',
      thumbnail: 'https://randomuser.me/api/portraits/thumb/men/24.jpg',
    },
    nat: 'AU',
  },
  {
    gender: 'male',
    name: {
      title: 'mr',
      first: 'paulo',
      last: 'seyfarth',
    },
    location: {
      street: 'talstra√üe 94',
      city: 'bogen',
      state: 'bayern',
      postcode: 74635,
      coordinates: {
        latitude: '18.2360',
        longitude: '72.7154',
      },
      timezone: {
        offset: '-3:00',
        description: 'Brazil, Buenos Aires, Georgetown',
      },
    },
    email: 'paulo.seyfarth@example.com',
    login: {
      uuid: 'af5ee15a-da22-4264-8c94-7cd57e21bf97',
      username: 'crazybutterfly740',
      password: 'colt45',
      salt: '7LDAWnJY',
      md5: 'e5bbaee291478ce51472ec82d31f97c6',
      sha1: 'ea8722d0fc4ba69dc43e36661468f5bda5199799',
      sha256: 'e31a953387293ab446bf702c20013a91f1c91793fe59fe425a6e6502d502ad50',
    },
    dob: {
      date: '1969-06-18T22:13:53Z',
      age: 49,
    },
    registered: {
      date: '2004-09-15T22:48:16Z',
      age: 13,
    },
    phone: '0477-2629824',
    cell: '0175-1061186',
    id: {
      name: '',
      value: null,
    },
    picture: {
      large: 'https://randomuser.me/api/portraits/men/98.jpg',
      medium: 'https://randomuser.me/api/portraits/med/men/98.jpg',
      thumbnail: 'https://randomuser.me/api/portraits/thumb/men/98.jpg',
    },
    nat: 'DE',
  },
  {
    gender: 'male',
    name: {
      title: 'mr',
      first: 'mathias',
      last: 'mortensen',
    },
    location: {
      street: '9165 fredensvej',
      city: 'pandrup',
      state: 'nordjylland',
      postcode: 39076,
      coordinates: {
        latitude: '-12.9436',
        longitude: '101.5652',
      },
      timezone: {
        offset: '+7:00',
        description: 'Bangkok, Hanoi, Jakarta',
      },
    },
    email: 'mathias.mortensen@example.com',
    login: {
      uuid: '39d341b7-2f7b-4b4a-bfd4-6921922da746',
      username: 'beautifulladybug715',
      password: 'carpente',
      salt: 'KpPfkc93',
      md5: '5f976528c474e2c1479464183d7800d9',
      sha1: '6d25ac15f23bf114e05cb940e7c61925b3274746',
      sha256: 'cf715b8dc50773efd203d755f2b0c268fd588e5b98204545ad1a71d30b66e4b2',
    },
    dob: {
      date: '1990-04-03T15:42:14Z',
      age: 28,
    },
    registered: {
      date: '2014-05-19T01:46:06Z',
      age: 4,
    },
    phone: '04223613',
    cell: '62185716',
    id: {
      name: 'CPR',
      value: '722632-7968',
    },
    picture: {
      large: 'https://randomuser.me/api/portraits/men/13.jpg',
      medium: 'https://randomuser.me/api/portraits/med/men/13.jpg',
      thumbnail: 'https://randomuser.me/api/portraits/thumb/men/13.jpg',
    },
    nat: 'DK',
  },
];

export const swapiPeople = [
  {
    name: 'Leia Organa',
    height: '150',
    mass: '49',
    hair_color: 'brown',
    skin_color: 'light',
    eye_color: 'brown',
    birth_year: '19BBY',
    gender: 'female',
    homeworld: 'https://swapi.co/api/planets/2/',
    films: [
      'https://swapi.co/api/films/2/',
      'https://swapi.co/api/films/6/',
      'https://swapi.co/api/films/3/',
      'https://swapi.co/api/films/1/',
      'https://swapi.co/api/films/7/',
    ],
    species: [
      'https://swapi.co/api/species/1/',
    ],
    vehicles: [
      'https://swapi.co/api/vehicles/30/',
    ],
    starships: [],
    created: '2014-12-10T15:20:09.791000Z',
    edited: '2014-12-20T21:17:50.315000Z',
    url: 'https://swapi.co/api/people/5/',
  },
  {
    name: 'Obi-Wan Kenobi',
    height: '182',
    mass: '77',
    hair_color: 'auburn, white',
    skin_color: 'fair',
    eye_color: 'blue-gray',
    birth_year: '57BBY',
    gender: 'male',
    homeworld: 'https://swapi.co/api/planets/20/',
    films: [
      'https://swapi.co/api/films/2/',
      'https://swapi.co/api/films/5/',
      'https://swapi.co/api/films/4/',
      'https://swapi.co/api/films/6/',
      'https://swapi.co/api/films/3/',
      'https://swapi.co/api/films/1/',
    ],
    species: [
      'https://swapi.co/api/species/1/',
    ],
    vehicles: [
      'https://swapi.co/api/vehicles/38/',
    ],
    starships: [
      'https://swapi.co/api/starships/48/',
      'https://swapi.co/api/starships/59/',
      'https://swapi.co/api/starships/64/',
      'https://swapi.co/api/starships/65/',
      'https://swapi.co/api/starships/74/',
    ],
    created: '2014-12-10T16:16:29.192000Z',
    edited: '2014-12-20T21:17:50.325000Z',
    url: 'https://swapi.co/api/people/10/',
  },
];

