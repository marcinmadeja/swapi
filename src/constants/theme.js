import { createMuiTheme } from 'material-ui/styles';
import blueGrey from 'material-ui/colors/blueGrey';
import yellow from 'material-ui/colors/yellow';

const templateStandard = {
  palette: {
    primary: blueGrey,
    secondary: yellow,
    borderColor: '#e6ecf5',
    background: {
      default: '#EDF2F6',
    },
  },
  borderRadius: '5px',
};

const mainTheme = createMuiTheme(templateStandard);

export const gutter = '15px';
export const containerWidth = '1300px';

export default mainTheme;
