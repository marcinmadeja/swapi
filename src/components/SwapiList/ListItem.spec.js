import React from 'react';
import 'jest-styled-components';
import { MemoryRouter } from 'react-router-dom';
import { swapiPeople } from 'constants/testConstants';
import ListItem, { createSeName } from './ListItem';

const selectedPerson = swapiPeople[0];

const defaultProps = {
  classes: {},
  item: {},
  ListDetails: () => { return null; },
  errors: false,
  listName: '',
};

const setup = (props = {}, children = null) => {
  props = { ...defaultProps, ...props };
  const component = <ListItem {...props}>{children}</ListItem>;
  const shallowComponent = shallow(component);
  const componentWithRouter = <MemoryRouter><ListItem {...props}>{children}</ListItem></MemoryRouter>;

  return {
    component,
    shallowComponent,
    componentWithRouter,
  };
};

describe('ListItem', () => {
  it('should render', () => {
    const { shallowComponent } = setup({ item: selectedPerson, listName: 'people' });
    expect(shallowComponent.exists()).toBe(true);
  });

  it('should match snapshot', () => {
    const { componentWithRouter } = setup({ item: selectedPerson, listName: 'people' });
    const tree = renderer
      .create(componentWithRouter)
      .toJSON();

    expect(tree).toMatchSnapshot(tree);
  });

  it('should create good ', () => {
    const seoName = createSeName(
      'Leia Organa',
      'https://swapi.co/api/people/5/',
      'people',
    );

    expect(seoName).toEqual('/people/leia-organa/5');
  });

  it('should create wrong link', () => {
    expect(createSeName(
      'Leia Organa',
      'https://swapi.co/api/people/7/',
      'people',
    )).not.toBe('/people/leia-organa/5');

    expect(createSeName(
      'Leia Organa',
      'https://swapi.co/api/people/5/',
      'peoplee',
    )).not.toBe('/people/leia-organa/5');
  });
});
