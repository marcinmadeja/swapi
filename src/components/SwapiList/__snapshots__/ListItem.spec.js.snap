// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[object Object] 1`] = `
<div
  className="MuiPaper-root-4 MuiPaper-elevation2-8 MuiPaper-rounded-5 ListItem-card-1"
>
  <div
    className="MuiCardContent-root-31 ListItem-cardContent-2"
  >
    <div
      className="MuiCardActions-root-32 ListItem-cardActions-3"
    >
      <button
        aria-label="Add to favorites"
        className="MuiButtonBase-root-40 MuiIconButton-root-34 MuiCardActions-action-33"
        disabled={false}
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiIconButton-label-39"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-42"
            color={undefined}
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-48"
        />
      </button>
      <a
        aria-label="Link"
        className="MuiButtonBase-root-40 MuiIconButton-root-34 MuiCardActions-action-33"
        href="/people/leia-organa/5"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex="0"
      >
        <span
          className="MuiIconButton-label-39"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-42"
            color={undefined}
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-48"
        />
      </a>
    </div>
  </div>
</div>
`;
